### PR TITLE
Add Property node to treat property keys as identifiers 

### PR DESCRIPTION
**Pull request by caithagoras**
Table/Schema properties were represented as String-to-Expression map,
which cannot handle delimited identifiers correctly as quotes are
treated as part of the property name.
Create SQL tree node Property to replace the maps used in CreateSchema,
 CreateTable, and CreateTableAsSelect.

### CODE REVIEW

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/TableProperties.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+    public List<TableProperty> getProperties()
+    {
+        return properties;
+    }
+
+    public Map<String, Expression> buildPropertyMap()
+    {
+        return properties.stream()
+                .collect(toImmutableMap(element -> element.getName().getValue(), TableProperty::getValue));
```
Comment (electrum) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/TableProperties.java]: This needs to lowercase the name. We can't do case sensitive until we change the SPI and pass identifiers to connectors.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/TableProperties.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+                .sorted(Comparator.comparing(property -> property.getName().getValue()))
+                .collect(toImmutableList());
+    }
+
+    public List<TableProperty> getProperties()
+    {
+        return properties;
+    }
+
+    public Map<String, Expression> buildPropertyMap()
```
Comment (martint) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/TableProperties.java]: This shouldn't be part of the node. It needs to take into account the rules for determining whether two identifiers are equivalent and raise an appropriate semantic error if there are duplicates. This logic belongs in the analysis phase.

Comment (caithagoras) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/TableProperties.java]: Yes, I have already do duplicate checking and throwing SemanticException so in StatementAnalayzer: https://github.com/prestodb/presto/pull/8791/files#diff-1a785e48f45c659076f1e4ec089517deR485

This is a helper method for places where a map is needed, since it has callers from different modules. (It is not used by StatementAnalyzer)

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/TableProperties.java (hunk 1 above)
Comment (caithagoras) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/TableProperties.java]: Thanks, I added that logic to the constructor of TableProperty such that name should only be lower-case identifiers.

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java [hunk 3]
```diff
@@ -262,7 +280,22 @@ public Node visitCreateTable(SqlBaseParser.CreateTableContext context)
         if (context.COMMENT() != null) {
             comment = Optional.of(((StringLiteral) visit(context.string())).getValue());
         }
-        return new CreateTable(getLocation(context), getQualifiedName(context.qualifiedName()), visit(context.tableElement(), TableElement.class), context.EXISTS() != null, processTableProperties(context.tableProperties()), comment);
+
+        TableProperties properties;
+        if (context.tableProperties() != null) {
```
Comment (kokosing) [inline presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java]: this `IF` block is repeated multiple times, can you please extract it as method?

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/TableProperties.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    {
+        this(Optional.of(location), properties);
+    }
+
+    private TableProperties(Optional<NodeLocation> location, List<TableProperty> properties)
+    {
+        super(location);
+        requireNonNull(properties, "properties is null");
+        this.properties = properties.stream()
+                .sorted(Comparator.comparing(property -> property.getName().getValue()))
```
Comment (kokosing) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/TableProperties.java]: why do you sort?

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/TableProperty.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+    public Expression getValue()
+    {
+        return value;
+    }
+
+    @Override
+    public <R, C> R accept(AstVisitor<R, C> visitor, C context)
+    {
+        return visitor.visitTableProperty(this, context);
```
Comment (kokosing) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/TableProperty.java]: please make sure to update ast visitors

File: presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java [hunk 6]
```diff
@@ -1327,84 +1328,89 @@ public void testCreateTableAsSelect()
...
+                new TableProperty(new Identifier("string"), new StringLiteral("bar")),
+                new TableProperty(new Identifier("long"), new LongLiteral("42")),
+                new TableProperty(
+                        new Identifier("computed"),
+                        new FunctionCall(QualifiedName.of("concat"), ImmutableList.of(new StringLiteral("ban"), new StringLiteral("ana")))),
+                new TableProperty(new Identifier("a"), new ArrayConstructor(ImmutableList.of(new StringLiteral("v1"), new StringLiteral("v2"))))));
 
         assertStatement("CREATE TABLE foo " +
-                        "WITH ( string = 'bar', long = 42, computed = 'ban' || 'ana', a  = ARRAY[ 'v1', 'v2' ] ) " +
+                        "WITH ( string = 'bar', long = 42, computed = 'ban' || 'ana', a = ARRAY[ 'v1', 'v2' ] ) " +
```
Comment (kokosing) [inline presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java]: what was changed here? Is it related? If no please extract as separate commit or drop.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 7]
```diff
@@ -389,6 +396,20 @@ protected Scope visitDelete(Delete node, Optional<Scope> scope)
         }
 
         @Override
+        protected Scope visitCreateSchema(CreateSchema node, Optional<Scope> scope)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: you moved and changed these methods in the same commit, can you please separate such changes to separate commits. It is much easier to follow that way.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 8]
```diff
@@ -465,43 +482,62 @@ protected Scope visitCreateView(CreateView node, Optional<Scope> scope)
...
+                    .map(TableProperty::getName)
+                    .map(Identifier::getValue)
+                    .collect(groupingBy(identity(), counting()))
+                    .entrySet()
+                    .stream()
+                    .filter(entry -> entry.getValue() > 1)
+                    .map(Entry::getKey)
+                    .findAny();
+            if (duplicateProperty.isPresent()) {
+                throw new SemanticException(DUPLICATE_TABLE_PROPERTIES, node, "Duplicate table property '%s'", duplicateProperty.get());
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: I would also have this change (raising this semantic exception) as separate commit. It seems to be logically separate to introduction of `TableProperties`

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/TableProperties.java (hunk 1 above)
Comment (martint) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/TableProperties.java]: No, that's no the right place to put it. The downcasing is a property of the analyzer, not the AST.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/TableProperties.java (hunk 2 above)
Comment (martint) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/TableProperties.java]: Those uses are potentially wrong, then, since they are making assumptions about how this method treats delimited identifiers.

Move the helper method to where it's used. The AST node is not the right place for it.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/TableProperties.java (hunk 4 above)
Comment (caithagoras) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/TableProperties.java]: Existing test expects properties fields are sorted.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/TableProperty.java (hunk 5 above)
Comment (caithagoras) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/TableProperty.java]: It has been updated.

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java (hunk 3 above)
Comment (caithagoras) [inline presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java]: Done. Although there are other places that repeats such null checking, not sure if I should follow.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 8 above)
Comment (caithagoras) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: The origin implementation creates a property-name-to-expression map during parsing, and duplication is de-duped silently, which is incorrectly. Separating the change will require rewrite the incorrect old behavior. I don't think that's desirable?

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/TableProperties.java (hunk 4 above)
Comment (caithagoras) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/TableProperties.java]: Removing the sorting

File: presto-parser/src/main/java/com/facebook/presto/sql/SqlFormatter.java [hunk 9]
```diff
@@ -831,22 +830,23 @@ protected Void visitCreateTable(CreateTable node, Integer indent)
...
         }
 
-        private void appendTableProperties(StringBuilder builder, Map<String, Expression> properties)
+        @Override
+        protected Void visitTableProperties(TableProperties properties, Integer indent)
         {
-            if (!properties.isEmpty()) {
+            if (!properties.getProperties().isEmpty()) {
                 builder.append("\nWITH (\n");
-                // Always output the table properties in sorted order
```
Comment (electrum) [inline presto-parser/src/main/java/com/facebook/presto/sql/SqlFormatter.java]: Why are you removing the sorting? This was done so that `SHOW CREATE TABLE` would be deterministic. Otherwise, that logic needs to go in every connector. Unit tests rely on this.

We might decide that the formatting should preserve the ordering returned by the connector, but if so, that should be a separate change. (and we'll need to update all the connectors and unit tests which rely on this)

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java [hunk 10]
```diff
@@ -262,18 +262,22 @@ public Node visitCreateTable(SqlBaseParser.CreateTableContext context)
...
+                getLocation(context),
+                getQualifiedName(context.qualifiedName()),
+                visit(context.tableElement(), TableElement.class),
+                context.EXISTS() != null,
+                processTableProperties(context.tableProperties()),
+                comment);
     }
 
-    private Map<String, Expression> processTableProperties(TablePropertiesContext tablePropertiesContext)
+    private TableProperties processTableProperties(@Nullable SqlBaseParser.TablePropertiesContext context)
```
Comment (electrum) [inline presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java]: Keep `TablePropertiesContext` static imported

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java [hunk 11]
```diff
@@ -262,18 +262,22 @@ public Node visitCreateTable(SqlBaseParser.CreateTableContext context)
...
     {
-        ImmutableMap.Builder<String, Expression> properties = ImmutableMap.builder();
-        if (tablePropertiesContext != null) {
-            for (TablePropertyContext tablePropertyContext : tablePropertiesContext.tableProperty()) {
-                properties.put(tablePropertyContext.identifier().getText(), (Expression) visit(tablePropertyContext.expression()));
-            }
+        if (context != null) {
```
Comment (electrum) [inline presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java]: This code is trivial and only used in one place. Let's inline it.

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java [hunk 12]
```diff
@@ -460,6 +464,21 @@ public Node visitDescribeInput(SqlBaseParser.DescribeInputContext context)
                 (Identifier) visit(context.identifier()));
     }
 
+    @Override
+    public Node visitTableProperties(SqlBaseParser.TablePropertiesContext context)
+    {
+        List<TableProperty> properties = context.tableProperty().stream()
+                .map(property -> (TableProperty) visit(property))
```
Comment (electrum) [inline presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java]: You can write this as
```java
.map(this::visit)
.map(TableProperty.class::cast)

Comment (electrum) [inline presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java]: Actually, use `visit()`:
```java
return new TableProperties(getLocation(context), visit(context.tableProperty(), TableProperty.class));

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/TableProperties.java [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static java.util.Objects.requireNonNull;
+
+public class TableProperties
```
Comment (electrum) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/TableProperties.java]: Why do we need a separate node for a list? I don't see the benefit over putting the list inline.

File: presto-parser/src/main/java/com/facebook/presto/sql/SqlFormatter.java [hunk 14]
```diff
@@ -831,22 +830,20 @@ protected Void visitCreateTable(CreateTable node, Integer indent)
...
-                String propertyList = ImmutableSortedMap.copyOf(properties).entrySet().stream()
-                        .map(entry -> INDENT + formatName(entry.getKey()) + " = " + formatExpression(entry.getValue(), parameters))
-                        .collect(joining(",\n"));
-                builder.append(propertyList);
-                builder.append("\n").append(")");
+            if (properties.isEmpty()) {
+                return "";
             }
+            String propertyList = properties.stream()
+                    .map(element -> formatExpression(element.getName(), parameters) + " = " + formatExpression(element.getValue(), parameters))
```
Comment (electrum) [inline presto-parser/src/main/java/com/facebook/presto/sql/SqlFormatter.java]: This changes the formatting and will break the tests (which are correct). Please preserve the original formatting.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 15]
```diff
@@ -503,6 +503,30 @@ protected Scope visitRenameSchema(RenameSchema node, Optional<Scope> scope)
...
+            validateTableProperties(node.getProperties(), scope);
+            return createAndAssignScope(node, scope);
+        }
+
+        private void validateTableProperties(List<TableProperty> properties, Optional<Scope> scope)
+        {
+            Set<String> propertyNames = new HashSet<>();
+            for (TableProperty property : properties) {
+                if (!propertyNames.add(property.getName().getValue())) {
+                    throw new SemanticException(DUPLICATE_SCHEMA_PROPERTY, property, "Duplicate table property '%s'", property.getName().getValue());
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: This is the wrong error code for table properties. Probably the easiest thing is to pass in the error code to use on failure.

File: presto-parser/src/main/java/com/facebook/presto/sql/SqlFormatter.java (hunk 14 above)
Comment (caithagoras) [reply presto-parser/src/main/java/com/facebook/presto/sql/SqlFormatter.java]: A separate PR has been merged to address the ordering. ShowQueryRewrites now sorts the properties rather than SqlFormatter.

**Comment by caithagoras** [comment]
I have merged all PRs that should go before this. Please take a look.

File: presto-main/src/main/java/com/facebook/presto/sql/NodeUtils.java [hunk 16]
```diff
@@ -28,4 +33,10 @@ private NodeUtils() {}
     {
         return orderBy.map(OrderBy::getSortItems).orElse(ImmutableList.of());
     }
+
+    public static Map<String, Expression> buildMapFromProperties(List<Property> properties)
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/NodeUtils.java]: No need for `build`, we can call this `mapFromProperties`.

File: presto-main/src/main/java/com/facebook/presto/sql/NodeUtils.java [hunk 17]
```diff
@@ -28,4 +33,10 @@ private NodeUtils() {}
...
+    {
+        return properties.stream()
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/NodeUtils.java]: I would format this like
```java
return properties.stream().collect(toImmutableMap(
        property -> property.getName().getValue(),
        Property::getValue);

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 18]
```diff
@@ -588,12 +600,16 @@ protected Scope visitCall(Call node, Optional<Scope> scope)
...
+        private void validateProperties(List<Property> properties, Optional<Scope> scope)
         {
-            for (Expression expression : properties.values()) {
-                // analyze property value expressions which must be constant
-                createConstantAnalyzer(metadata, session, analysis.getParameters(), analysis.isDescribe())
-                        .analyze(expression, createScope(scope));
+            Set<String> propertyNames = new HashSet<>();
+            for (Property property : properties) {
+                if (!propertyNames.add(property.getName().getValue())) {
+                    throw new SemanticException(DUPLICATE_PROPERTY, property, "Duplicate property '%s'", property.getName().getValue());
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: It's better to put the value at the end rather than delimiting it:
```java
"Duplicate property: %s"

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 19]
```diff
@@ -505,6 +508,15 @@ protected Scope visitCreateTable(CreateTable node, Optional<Scope> scope)
         }
 
         @Override
+        protected Scope visitProperty(Property node, Optional<Scope> scope)
+        {
+            // Table property value expressions must be constant
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Remove "Table" here since this is used for schema as well

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java [hunk 20]
```diff
@@ -487,11 +488,15 @@ protected Node visitShowCreate(ShowCreate node, Void context)
                     sqlProperties.put(propertyName, sqlExpression);
                 }
 
+                List<Property> propertyNodes = sqlProperties.build().entrySet().stream()
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java]: I think this can be named just `properties`

File: presto-parser/src/main/java/com/facebook/presto/sql/SqlFormatter.java [hunk 21]
```diff
@@ -720,15 +720,14 @@ protected Void visitDelete(Delete node, Integer context)
         }
 
         @Override
-        protected Void visitCreateSchema(CreateSchema node, Integer context)
+        protected Void visitCreateSchema(CreateSchema node, Integer indent)
```
Comment (electrum) [inline presto-parser/src/main/java/com/facebook/presto/sql/SqlFormatter.java]: Don't rename the parameter. It's unused and this rename is unrelated to this commit.

File: presto-parser/src/main/java/com/facebook/presto/sql/SqlFormatter.java [hunk 22]
```diff
@@ -830,21 +829,20 @@ protected Void visitCreateTable(CreateTable node, Integer indent)
...
-                String propertyList = properties.entrySet().stream()
-                        .map(entry -> INDENT + formatName(entry.getKey()) + " = " + formatExpression(entry.getValue(), parameters))
-                        .collect(joining(",\n"));
-                builder.append(propertyList);
-                builder.append("\n").append(")");
+            if (properties.isEmpty()) {
+                return "";
             }
+            String propertyList = properties.stream()
+                    .map(element -> INDENT + formatExpression(element.getName(), parameters) + " = " + formatExpression(element.getValue(), parameters))
```
Comment (electrum) [inline presto-parser/src/main/java/com/facebook/presto/sql/SqlFormatter.java]: Let's add a line break
```java
.map(element -> INDENT +
        formatExpression(element.getName(), parameters) + " = " +
        formatExpression(element.getValue(), parameters))

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java [hunk 23]
```diff
@@ -262,26 +269,19 @@ public Node visitCreateTable(SqlBaseParser.CreateTableContext context)
...
                 getLocation(context),
                 getQualifiedName(context.qualifiedName()),
                 visit(context.tableElement(), TableElement.class),
                 context.EXISTS() != null,
-                processProperties(context.properties()),
+                properties,
                 comment);
     }
 
-    private Map<String, Expression> processProperties(PropertiesContext propertiesContext)
```
Comment (electrum) [inline presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java]: We should be able to keep this method and have it return `List<Property>`. This avoids duplicating the code. You might be able to use the existing `visit()` helper.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/CreateTableAsSelect.java [hunk 24]
```diff
@@ -30,28 +28,28 @@
...
 
-    private CreateTableAsSelect(Optional<NodeLocation> location, QualifiedName name, Query query, boolean notExists, Map<String, Expression> properties, boolean withData, Optional<List<Identifier>> columnAliases, Optional<String> comment)
+    private CreateTableAsSelect(Optional<NodeLocation> location, QualifiedName name, Query query, boolean notExists, List<Property> properties, boolean withData, Optional<List<Identifier>> columnAliases, Optional<String> comment)
     {
         super(location);
         this.name = requireNonNull(name, "name is null");
         this.query = requireNonNull(query, "query is null");
         this.notExists = notExists;
-        this.properties = ImmutableMap.copyOf(requireNonNull(properties, "properties is null"));
+        this.properties = requireNonNull(properties, "properties is null");
```
Comment (electrum) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/CreateTableAsSelect.java]: This should make a defensive copy of the list.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/CreateSchema.java [hunk 25]
```diff
@@ -29,24 +27,24 @@
...
     }
 
-    private CreateSchema(Optional<NodeLocation> location, QualifiedName schemaName, boolean notExists, Map<String, Expression> properties)
+    private CreateSchema(Optional<NodeLocation> location, QualifiedName schemaName, boolean notExists, List<Property> properties)
     {
         super(location);
         this.schemaName = requireNonNull(schemaName, "schemaName is null");
         this.notExists = notExists;
-        this.properties = ImmutableMap.copyOf(requireNonNull(properties, "properties is null"));
+        this.properties = requireNonNull(properties, "properties is null");
```
Comment (electrum) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/CreateSchema.java]: This should make a defensive copy of the list.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/CreateSchema.java [hunk 26]
```diff
@@ -73,9 +71,7 @@ public boolean isNotExists()
     @Override
     public List<Node> getChildren()
     {
-        ImmutableList.Builder<Node> nodes = ImmutableList.builder();
-        nodes.addAll(properties.values());
-        return nodes.build();
+        return ImmutableList.copyOf(properties);
```
Comment (electrum) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/CreateSchema.java]: No need to copy here (the class will store an immutable list)

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/Property.java [hunk 27]
```diff
@@ -0,0 +1,97 @@
...
+    public boolean equals(Object obj)
+    {
+        if (this == obj) {
+            return true;
+        }
+        if (obj == null || getClass() != obj.getClass()) {
+            return false;
+        }
+        Property other = (Property) obj;
+        return Objects.equals(name, other.name) && Objects.equals(value, other.value);
```
Comment (electrum) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/Property.java]: Add line break after `&&`

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/Property.java [hunk 28]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    @Override
+    public List<? extends Node> getChildren()
+    {
+        return ImmutableList.of(name, value);
+    }
+
+    @Override
+    public String toString()
```
Comment (electrum) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/Property.java]: Put the `toString` method at the end

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java (hunk 20 above)
Comment (caithagoras) [reply presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java]: There is already a local variable named `properties`

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java (hunk 23 above)
Comment (caithagoras) [reply presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java]: There is not much duplicating logic. If you're referring to `if (context.properties() != null)`, this is not extracted elsewhere in the file.

**Comment by caithagoras** [comment]
ping for review @electrum
